### ISSUE TITLE
Development Roadmap 

### ISSUE THREAD

**Issue by Ying1123** [sgl-project/sglang#157]
## Function Calling
- Frontend
    - Add `tools` argument in `sgl.gen`. See also guidance [tools](https://github.com/guidance-ai/guidance/blob/d1bbe1c698cbb201f89556d71193993e78c0686b/README.md?plain=1#L102)
- Backend
    - OpenAI: Translate to their function calling API (https://platform.openai.com/docs/guides/function-calling).
      - #573 
    - Local Models (SGLang)
        1. Use SGLang primitives (regex, select) and constrained decoding to implement a workflow 
        2. Directly use models that support function calling (e.g., Gorilla OpenFunctions, https://huggingface.co/jondurbin/bagel-dpo-7b-v0.4#prompting-strategies)
     - Local Models (OpenAI-compatible API)

## High-level Pythonic Interface
 - #39
 
## Inference Optimizations
- Speculative decoding for local models
- Speculative execution for OpenAI Chat API
  - #48 

## Structured Decoding
- Support parallel JSON decoding https://github.com/varunshenoy/super-json-mode/issues/8
- Support auto parallel decoding  https://arxiv.org/abs/2401.06761

## Compiler
- Support tracing and compiling `sgl.fork` 
- Support sending a full serialized SGL program to the server

## LoRA Support
- Port multi-LoRA batching and unified memory from S-LoRA

## Model Coverage
- Vision Langauge Models. Support top-performing models from https://github.com/open-compass/VLMEvalKit
- Language Models. Port the implementation of popular models from https://github.com/vllm-project/vllm/tree/main/vllm/model_executor/models. ([help](https://github.com/sgl-project/sglang/blob/main/docs/model_support.md))
## Device Coverage
- AMD support. Investigate AMD support in Trion and FlashInfer.
- CPU support. This is better done by adding a llama.cpp backend.

**Comment by AriMKatz** [comment]
Are there still plans for a high level pythonic interface? https://github.com/sgl-project/sglang/issues/39#issuecomment-1899351565

**Comment by Ying1123** [comment]
> Are there still plans for a high level pythonic interface? [#39 (comment)](https://github.com/sgl-project/sglang/issues/39#issuecomment-1899351565)

Hi @AriMKatz, thanks for the reference. This is very important, I just added it.

**Comment by nivibilla** [comment]
For the vision models support, is it possible to align with the openai gpt4v API? 
https://platform.openai.com/docs/guides/vision

**Comment by aliencaocao** [comment]
Are there plans for loading models in 8bit or 4bit?

**Comment by Ying1123** [comment]
> For the vision models support, is it possible to align with the openai gpt4v API? https://platform.openai.com/docs/guides/vision

@nivibilla Yes, it is already aligned with the openai gpt4v API, see [here](https://github.com/sgl-project/sglang/blob/4d303c4fa365dbe8b4d474be6e613954bb829939/test/srt/test_openai_server.py#L98).
You can also find a runnable example of serving it with Sky Serve [here](https://github.com/skypilot-org/skypilot/pull/3145).

**Comment by Ying1123** [comment]
> Are there plans for loading models in 8bit or 4bit?

@aliencaocao Thanks for the question! The AWQ and GPTQ are already supported. But we do not support an automatic dtype translation yet. You are welcome to submit a PR for that.

**Comment by aliencaocao** [comment]
> > Are there plans for loading models in 8bit or 4bit?
> 
> @aliencaocao Thanks for the question! The AWQ and GPTQ are already supported. But we do not support an automatic dtype translation yet. You are welcome to submit a PR for that.

I'm looking to load llava 1.6 in 8bit, but it does not seem that llava series has AWQ or GPTQ quants, or did I miss out anything here?

EDIT: I saw 1.5 has but not 1.6 yet. Perhaps its just too new and no one did a calibration yet.

**Comment by jackdiy** [comment]
have considered A plan for MEMGPT ？ improve memory and long context

**Comment by qeternity** [comment]
Hi all - is anyone working on the S-LoRA integration currently? I see the branch, but it looks a few months old.

Would love to see this, happy to pick up from existing work or start fresh.

**Comment by Ying1123** [comment]
> Hi all - is anyone working on the S-LoRA integration currently? I see the branch, but it looks a few months old.
> 
> Would love to see this, happy to pick up from existing work or start fresh.

Hi @qeternity, I was working on it but have been blocked by other affairs. You are welcome to contribute, either continue on the branch or start fresh! I'll be happy to review and collaborate.

**Comment by Bit0r** [comment]
Tools support is very important, which is necessary for many use cases.

**Comment by omri-sap** [comment]
Is TinyLlama supported? [TinyLlama/TinyLlama-1.1B-Chat-v1.0](https://huggingface.co/TinyLlama/TinyLlama-1.1B-Chat-v1.0)
generation seems a bit slow...

**Comment by wille-42** [comment]
I see llama.cpp integration is on the roadmap. When will this feature be delivered? It would be very nice to have it , since it will support running local LLMs, such as llama models, on Mac computers and experiment them with the powerful and expressive SGLang.

**Comment by gintasz** [comment]
I'd request to include support for Phi-3-mini

**Comment by binarycrayon** [comment]
> > Hi all - is anyone working on the S-LoRA integration currently? I see the branch, but it looks a few months old.
> > Would love to see this, happy to pick up from existing work or start fresh.
> 
> Hi @qeternity, I was working on it but have been blocked by other affairs. You are welcome to contribute, either continue on the branch or start fresh! I'll be happy to review and collaborate.

Hi which branch is it? looks like better start fresh

**Comment by zhyncs** [comment]
I can help by getting rid of the vLLM in the dependencies.

**Comment by Ying1123** [comment]
Moved to https://github.com/sgl-project/sglang/issues/634

**Comment by CedricHwong** [comment]
Is the “tools” ok now？
